File: msr/reference/go/matrix-ops.go
package main

func mat_add(a [][]int, b [][]int) [][]int {
	out := [][]int{}
	for i := range a {
		row := []int{}
		for j := range a[i] {
			row = append(row, a[i][j]+b[i][j])
		}
		out = append(out, row)
	}
	return out
}

func mat_transpose(m [][]int) [][]int {
	out := [][]int{}
	if len(m) == 0 {
		return out
	}
	for j := 0; j < len(m[0]); j++ {
		col := []int{}
		for i := range m {
			col = append(col, m[i][j])
		}
		out = append(out, col)
	}
	return out
}

func mat_scale(m [][]int, s int) [][]int {
	out := [][]int{}
	for _, r := range m {
		row := []int{}
		for _, x := range r {
			row = append(row, x*s)
		}
		out = append(out, row)
	}
	return out
}

func mat_dot_row(row_a []int, col_b []int) int {
	total := 0
	for i := range row_a {
		total += row_a[i] * col_b[i]
	}
	return total
}

File: msr/reference/zig/matrix-ops.zig
const std = @import("std");

pub fn mat_add(allocator: std.mem.Allocator, a: []const []const i64, b: []const []const i64) ![][]i64 {
    const out = try allocator.alloc([]i64, a.len);
    for (a, b, 0..) |ra, rb, i| {
        out[i] = try allocator.alloc(i64, ra.len);
        for (ra, rb, 0..) |x, y, j| out[i][j] = x + y;
    }
    return out;
}

pub fn mat_transpose(allocator: std.mem.Allocator, m: []const []const i64) ![][]i64 {
    const ncols = if (m.len == 0) 0 else m[0].len;
    const out = try allocator.alloc([]i64, ncols);
    for (0..ncols) |j| {
        out[j] = try allocator.alloc(i64, m.len);
        for (m, 0..) |row, i| out[j][i] = row[j];
    }
    return out;
}

pub fn mat_scale(allocator: std.mem.Allocator, m: []const []const i64, s: i64) ![][]i64 {
    const out = try allocator.alloc([]i64, m.len);
    for (m, 0..) |row, i| {
        out[i] = try allocator.alloc(i64, row.len);
        for (row, 0..) |x, j| out[i][j] = x * s;
    }
    return out;
}

pub fn mat_dot_row(row_a: []const i64, col_b: []const i64) i64 {
    var total: i64 = 0;
    for (row_a, col_b) |x, y| total += x * y;
    return total;
}
